### ISSUE TITLE
Search Mechanism 

### ISSUE THREAD

**Issue by AliOsm** [emad-elsaid/rubyfunctions#22]
# Problem

- The users can't search for a function in the website.

# Suggested solution

- Implement a simple search mechanism based on `usage` field keywords.

**Comment by emad-elsaid** [comment]
Yeah that's a really hard one to get right, full text search is usually an expensive feature to implement, either we'll have to put the load on postgres which is awful performance for full text search, or elastic search which is a eats up all memory, or something custom.

so you have a clean way to do this without making a lot of mess? because it's very important to have actually

**Comment by AliOsm** [comment]
If it is very expensive as you mentioned, we can do one of the following:

- Extract keywords from function usage field, and run the search on top of them (this can be used to enhance the tags feature).
- Rely on the tags field.

The first option comes with the usage of some of machine learning techniques to extract the keywords, while the second one will rely on how good the users will fill the tags field.
